gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1503 @@
+	SetConfigOption("arenadata_toolkit.tracking_schemas",
+					new_schemas ? new_schemas : "",
+					PGC_SUSET, PGC_S_DATABASE);
+
+	if (current_schemas)
+		pfree(current_schemas);
+	if (new_schemas)
+		pfree(new_schemas);
+}
+
+Datum
+tracking_register_schema(PG_FUNCTION_ARGS)
+{
+	const char *schema_name = NameStr(*PG_GETARG_NAME(0));
+	Oid			dbid = get_dbid(PG_GETARG_OID(1));
+
+	if (Gp_role != GP_ROLE_DISPATCH)
+		ereport(ERROR,
+				(errmsg("Cannot execute tracking_register_schema outside query dispatcher")));
+
+	if (schema_name == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("schema does not exist")));
+
+	if (!SearchSysCacheExists1(NAMESPACENAME, CStringGetDatum(schema_name)))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("schema %s does not exist", schema_name)));
+
+	elog(LOG, "[arenadata_toolkit] registering schema %s in database %u for tracking", schema_name, dbid);
+
+	track_schema(schema_name, dbid, true);
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_unregister_schema(PG_FUNCTION_ARGS)
+{
+	const char *schema_name = NameStr(*PG_GETARG_NAME(0));
+	Oid			dbid = get_dbid(PG_GETARG_OID(1));
+
+	if (Gp_role != GP_ROLE_DISPATCH)
+	{
+		ereport(ERROR,
+				(errmsg("Cannot execute tracking_unregister_schema outside query dispatcher")));
+	}
+
+	if (schema_name == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("schema does not exist")));
+
+	if (!SearchSysCacheExists1(NAMESPACENAME, CStringGetDatum(schema_name)))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("schema with OID %s does not exist", schema_name)));
+
+	elog(LOG, "[arenadata_toolkit] registering schema %s in database %u for tracking", schema_name, dbid);
+
+	track_schema(schema_name, dbid, false);
+
+	PG_RETURN_BOOL(true);
+}
+
+static bool
+is_valid_relkind(char relkind)
+{
+	switch (relkind)
+	{
+		case RELKIND_RELATION:
+		case RELKIND_INDEX:
+		case RELKIND_SEQUENCE:
+		case RELKIND_TOASTVALUE:
+		case RELKIND_VIEW:
+		case RELKIND_COMPOSITE_TYPE:
+		case RELKIND_FOREIGN_TABLE:
+		case RELKIND_UNCATALOGED:
+		case RELKIND_MATVIEW:
+		case RELKIND_AOSEGMENTS:
+		case RELKIND_AOBLOCKDIR:
+		case RELKIND_AOVISIMAP:
+			return true;
+		default:
+			return false;
+	}
+}
+
+Datum
+tracking_set_relkinds(PG_FUNCTION_ARGS)
+{
+	char	   *relkinds_str = NameStr(*PG_GETARG_NAME(0));
+	Oid			dbid = get_dbid(PG_GETARG_OID(1));
+	char	   *token;
+	char	   *str_copy;
+	bool		seen_relkinds[256] = {false};
+	StringInfoData buf;
+	AlterDatabaseSetStmt stmt;
+	VariableSetStmt v_stmt;
+	A_Const		arg;
+
+	if (Gp_role != GP_ROLE_DISPATCH)
+	{
+		ereport(ERROR,
+				(errmsg("Cannot execute tracking_set_relkinds outside query dispatcher")));
+	}
+
+	if (relkinds_str == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid relkind set"),
+				 errhint("Valid relkinds are: 'r', 'i', 'S', 't', 'v', 'c', 'f', 'u', 'm', 'o', 'b', 'M'")));
```

Comment:
the string for 'Valid relkind' and 'Valid relstorages' are duplicated several times. I think better to make a single define (or a const string) for each such message and use it throughout the file.